// built-extension-sdk/controller/application-user.js
module.exports = {
	before: function(req, res){
		req.bclass = getAppUserClass(req.builtApp)
		return utils.Promise.resolve()
	},
	GET: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Query(req.payload.object)
			return utils.Promise.resolve()
		},
		_pre_send: function(req, res){
			return this.resSuccess(req, res, req.bobjekt.data)
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		},
		_post_send: function(req, res){
			return this.resSuccess(req, res, req.bobjekt)
		}
	},
	DELETE: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Query(req.payload.object)
			return utils.Promise.resolve()
		},
		_pre_send: function(req, res){
			return this.resSuccess(req, res, req.bobjekt.data)
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		},
		_post_send: function(req, res){
			return this.resSuccess(req, res, null)
		}
	},
	POST: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Object(req.payload.object)
			return utils.Promise.resolve()
		},
		_pre_send: function(req, res){
			return this.resSuccess(req, res, req.bobjekt.delta)
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		},
		_post_send: function(req, res){
			return this.resSuccess(req, res, req.bobjekt)
		}
	},
	'/:userUid': {
		GET:{
			_pre: function(req, res){
				req.bobjekt = req.bclass.Object(req.params.userUid)
				return utils.Promise.resolve()
			},
			_pre_send: function(req, res){
				return this.resSuccess(req, res, null)
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			},
			_post_send: function(req, res){
				return this.resSuccess(req, res, req.bobjekt)
			}
		},
		PUT: {
			_pre: function(req, res){
				req.bobjekt = req.bclass.Object(req.params.userUid).assign(req.payload.object)
				return utils.Promise.resolve()
			},
			_pre_send: function(req, res){
				return this.resSuccess(req, res, req.bobjekt.delta)
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			},
			_post_send: function(req, res){
				return this.resSuccess(req, res, req.bobjekt)
			}
		},
		DELETE: {
			_pre: function(req, res){
				req.bobjekt = req.bclass.Object(req.params.userUid)
				return utils.Promise.resolve()
			},
			_pre_send: function(req, res){
				return this.resSuccess(req, res, null)
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			},
			_post_send: function(req, res){
				return this.resSuccess(req, res, null)
			}
		}
	}
}

function getAppUserClass(bapp) {
  return bapp.Class('built_io_application_user')
}
