// built-extension-sdk/extension.js
var router        = require('./router')
var FrameworkApp  = require('./framework/app')
var RegisterHooks = require('./register-hooks')
var middlewares   = require('./middlewares')
var coreRoutes 		= require('./routes')
var express 			= require('express')
/**
 * constants
 */
// const MIDDLEWARE_PATH = /^\/(?:_pre|_post)\/v[0-9]\/.*/i

var ExtensionSDK = function(app, configs) {
	this.app          = app
	this.configs      = setConfig(configs)
	this.frameworkApp = FrameworkApp()
}

var setConfig = function(config){
	if(!config)
		return {}

	if(utils.isString(config)){
		return {
			extension_key: config
		}
	}
	if(utils.isPlainObject(config)){
		return config
	}
	checkTypeError(false, "config", "object")
}


var checkTypeError = function(isValid, value, expectedType){
	if(!isValid)
		throw new Error(value +" should be of type "+ expectedType)
}

ExtensionSDK.prototype.getAppInstance = function(){
	return this.app
}

ExtensionSDK.prototype.getConfig = function(){
	return this.configs
}

/**
 * Helper method for getting extension key.
 */
ExtensionSDK.prototype.getExtensionKey = function(){
	return this.configs.extension_key
}

ExtensionSDK.prototype.start = function(port){
	var that              = this
	var appConfig         = this.app.getConfig()

	var pluginRoutes      = this.configs.routes || {}
	var static            = this.configs.static
	var extensionKey      = this.configs.extension_key || 'blt_ext_key'
	var serverAppInstance = this.frameworkApp.getAppInstance()
	var appInstance       = this.getAppInstance()
	var applicationApiKey = appConfig.application_api_key
	var masterKey         = appConfig.master_key
	
	port                  = port || serverAppInstance.config.get('server:port')

	// We load user specific routes
	return utils.Promise.resolve()
	.then(function(){
		return middlewares.load("/api", [
      middlewares.consoleLogger,
      middlewares.reqParams,
      middlewares.pragmaticRest,
      middlewares.responseTime,
      middlewares.responseTimeout,
      middlewares.wrapInBuilt,
      middlewares.wrapInBuiltApp
    ], serverAppInstance, that)
	})
	.then(function(){
		return that.frameworkApp.startWithPort(port)
	})
	.then(function(){
		// var clonedRoutes = utils.clone(coreRoutes)
		return router(serverAppInstance, coreRoutes, pluginRoutes)
	})
	.then(function(){
		return RegisterHooks(that)
	})
	.then(function(){
		if(static)
			serverAppInstance.use(express.static(static))
	})
	.then(function(){
		return that.frameworkApp
	})
	.catch(function(err){
		if(err.stack)
			console.log("Error stack", err.stack)
		else
			console.log("Error", err)
	})
}

ExtensionSDK.prototype.stop = function(){
	return this.frameworkApp.stop()
}

module.exports = function(configs){
	return new ExtensionSDK(this, configs)
}
